### PR TITLE
psi-plus: add more build options

### PR DESCRIPTION
**Pull request by unclechu**
###### Motivation for this change

Psi+ itself supports many cmake build options. For instance “chat type” is a very common one that people customize for themselves. But on the Nix side there is no native representation of those options. It would be a lot nicer to have simple Nix-side build options. This MR adds Nix-native customization options for these:

1. `CHAT_TYPE`
2. Customizable `ENABLE_PLUGINS` option
3. `BUILD_PSIMEDIA`
4. `extraCmakeFlags` that only adds additional cmake flags to what we already have. In case a user doesn’t wanna override original value completely.

I tested builds with all 3 possible values of “chat type”. And I also tested other options as well. All works fine.

###### Things done

- [x] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [21.11 Release Notes (or backporting 21.05 Relase notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2111-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: pkgs/applications/networking/instant-messengers/psi-plus/default.nix
Comment (SuperSandro2000) [inline pkgs/applications/networking/instant-messengers/psi-plus/default.nix:55]:
```suggestion
  nativeBuildInputs = [ cmake qttools ]
    ++ lib.optional enablePsiMedia pkg-config;
```

File: pkgs/applications/networking/instant-messengers/psi-plus/default.nix
Comment (SuperSandro2000) [inline pkgs/applications/networking/instant-messengers/psi-plus/default.nix:69]:
```suggestion
  buildInputs = [
    qtbase qtmultimedia qtx11extras
    libidn qca-qt5 libXScrnSaver hunspell
    libsecret libgcrypt libotr html-tidy libgpgerror libsignal-protocol-c
    usrsctp
  ] ++ lib.optionals voiceMessagesSupport [
    gst_all_1.gst-plugins-base
    gst_all_1.gst-plugins-good
  ] ++ lib.optional (chatType == "webkit") qtwebkit
  ++ lib.optional (chatType == "webengine") qtwebengine;
```

Comment (unclechu) [reply pkgs/applications/networking/instant-messengers/psi-plus/default.nix:69]: @SuperSandro2000 are you following some style guide or is it a personal preference?

My original intention was to use a block indentation and every entry on its own line. If you take the last line and hide others ones:

``` nix
buildInputs = [
# ...
++ lib.optional (chatType == "webengine") qtwebengine;
```

They are on the same indentation level which looks a bit weird to me.

File: pkgs/applications/networking/instant-messengers/psi-plus/default.nix
Comment (unclechu) [reply pkgs/applications/networking/instant-messengers/psi-plus/default.nix:55]: @SuperSandro2000 if I compare this suggestion with another one:

``` nix
buildInputs = [
# ...
++ lib.optional (chatType == "webengine") qtwebengine;
```

``` nix
nativeBuildInputs = [ cmake qttools ]
  ++ lib.optional enablePsiMedia pkg-config;
```

They have different indentation level for their last lines. What is the intention of these suggestions?

Comment (SuperSandro2000) [reply pkgs/applications/networking/instant-messengers/psi-plus/default.nix:55]: I don't know 🤔 I would just use nixpkgs-fmt and use that.

File: pkgs/applications/networking/instant-messengers/psi-plus/default.nix
Comment (SuperSandro2000) [reply pkgs/applications/networking/instant-messengers/psi-plus/default.nix:69]: Yeah, maybe it needs to be intended, too.

File: pkgs/applications/networking/instant-messengers/psi-plus/default.nix
Comment (unclechu) [reply pkgs/applications/networking/instant-messengers/psi-plus/default.nix:55]: @SuperSandro2000 if I apply `nixpkgs-fmt` on this file these line are staying unaffected. I hope that’s okay that I decline these suggestions. They only problem with them is that they are inconsistent with each other, I couldn’t put some logic behind them.

File: pkgs/applications/networking/instant-messengers/psi-plus/default.nix
Comment (unclechu) [reply pkgs/applications/networking/instant-messengers/psi-plus/default.nix:69]: I would decline this one. In current form it’s being patch-friendly. Any new entry is added independently.

If you take this one even with indentation:
``` nix
buildInputs = [
  # ...
] ++ lib.optional (chatType == "webkit") qtwebkit
  ++ lib.optional (chatType == "webengine") qtwebengine;
```

Imagine adding another list:

``` nix
buildInputs = [
  # ...
] ++ lib.optional (chatType == "webkit") qtwebkit
  ++ lib.optional (chatType == "webengine") qtwebengine
  ++ lib.optionals foo [
  # ...
];
```

This would look just weird. But with original style it looks more or less okay:

``` nix
buildInputs =
  [
    # ...
  ] 
  ++ lib.optional (chatType == "webkit") qtwebkit
  ++ lib.optional (chatType == "webengine") qtwebengine
  ++ lib.optionals foo [
    # ...
  ];
```

**Comment by misuzu** [comment]
@unclechu I see you interested in working on `psi-plus`. Would you mind adding yourself as maintainer too?

**Review by misuzu — APPROVED** [review]
Result of `nixpkgs-review pr 129710` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package built:</summary>
  <ul>
    <li>psi-plus</li>
  </ul>
</details>

**Comment by unclechu** [comment]
> @unclechu I see you interested in working on `psi-plus`. Would you mind adding yourself as maintainer too?

Yeah, I was thinking of it too. Will add myself to the list of maintainers in another MR. But I have a question:

If I am a maintainer and I make a change, who reviews&approves that change?

File: pkgs/applications/networking/instant-messengers/psi-plus/default.nix
Comment (SuperSandro2000) [reply pkgs/applications/networking/instant-messengers/psi-plus/default.nix:55]: nixpkgs-fmt does not enforce the new lines to prevent to many changes. I will not remember this conversation and the next time anything changes in this file I will probably come up with the formatting again. Just follow it, it is used everywhere in nixpkgs and does not cause unnecessary merge conflicts.

Comment (unclechu) [reply pkgs/applications/networking/instant-messengers/psi-plus/default.nix:55]: @SuperSandro2000 sorry, I’m missing the point. Here is the diff that `nixpkgs-fmt` against that file gives me:
``` diff
diff --git a/pkgs/applications/networking/instant-messengers/psi-plus/default.nix b/pkgs/applications/networking/instant-messengers/psi-plus/default.nix
index a6a27c7fc0f..14f5a3a1c98 100644
--- a/pkgs/applications/networking/instant-messengers/psi-plus/default.nix
+++ b/pkgs/applications/networking/instant-messengers/psi-plus/default.nix
@@ -1,7 +1,21 @@
-{ lib, mkDerivation, fetchFromGitHub, cmake
-, qtbase, qtmultimedia, qtx11extras, qttools
-, libidn, qca-qt5, libXScrnSaver, hunspell
-, libsecret, libgcrypt, libotr, html-tidy, libgpgerror, libsignal-protocol-c
+{ lib
+, mkDerivation
+, fetchFromGitHub
+, cmake
+, qtbase
+, qtmultimedia
+, qtx11extras
+, qttools
+, libidn
+, qca-qt5
+, libXScrnSaver
+, hunspell
+, libsecret
+, libgcrypt
+, libotr
+, html-tidy
+, libgpgerror
+, libsignal-protocol-c
 , usrsctp
 
 , chatType ? "basic" # See the assertion below for available options
@@ -10,19 +24,19 @@
 
 , enablePlugins ? true
 
-# Voice messages
+  # Voice messages
 , voiceMessagesSupport ? true
 , gst_all_1
 
 , enablePsiMedia ? false
 , pkg-config
 
-, extraCmakeFlags ? [] # In addition to existing ones
+, extraCmakeFlags ? [ ] # In addition to existing ones
 }:
 
 assert builtins.elem (lib.toLower chatType) [
-  "basic"     # Basic implementation, no web stuff involved
-  "webkit"    # Legacy one, based on WebKit (see https://wiki.qt.io/Qt_WebKit)
+  "basic" # Basic implementation, no web stuff involved
+  "webkit" # Legacy one, based on WebKit (see https://wiki.qt.io/Qt_WebKit)
   "webengine" # QtWebEngine (see https://wiki.qt.io/QtWebEngine)
 ];
 
@@ -56,9 +70,19 @@ mkDerivation rec {
 
   buildInputs =
     [
-      qtbase qtmultimedia qtx11extras
-      libidn qca-qt5 libXScrnSaver hunspell
-      libsecret libgcrypt libotr html-tidy libgpgerror libsignal-protocol-c
+      qtbase
+      qtmultimedia
+      qtx11extras
+      libidn
+      qca-qt5
+      libXScrnSaver
+      hunspell
+      libsecret
+      libgcrypt
+      libotr
+      html-tidy
+      libgpgerror
+      libsignal-protocol-c
       usrsctp
     ]
     ++ lib.optionals voiceMessagesSupport [
```

Comment (unclechu) [reply pkgs/applications/networking/instant-messengers/psi-plus/default.nix:55]: @SuperSandro2000 
1. `nixpkgs-fmt` does not change the style of indentation
2. `nixpkgs-fmt` touches other stuff as you can see
3. > Just follow it, it is used everywhere in nixpkgs and does not cause unnecessary merge conflicts.

   The style I used also targeted to avoid merge conflicts. To be patch friendly. I showed you some examples that are not obvious how to approach. For instance that one with the lack of indentation before `++`, there would be a merge conflicts when a change comes.

   ``` nix
   buildInputs = [
   # ...
   ++ lib.optional (chatType == "webengine") qtwebengine;
   ```

   Consider this if it was fixed and another entry was added:

   ``` nix
   buildInputs = [
     # ...
   ] ++ lib.optional (chatType == "webkit") qtwebkit
     ++ lib.optional (chatType == "webengine") qtwebengine
     ++ lib.optionals foo [
     # ...
   ];
   ```

   Imagine the last record appeared the latest. Isn’t this looks weird?

4. If you show me the logic, the standard, I will follow. What I see now is 2 suggestions that are in conflicts with each other as they are using different indentation style.

Comment (unclechu) [reply pkgs/applications/networking/instant-messengers/psi-plus/default.nix]: @SuperSandro2000 
1. I changed that style to one that you proposed and I also replaced all `lib.optional` to `lib.optionals` even if it’s only one element inside, thus there are no more weird indentation cases, and also less potential merge conflicts in the future
2. I also applied `nixpkgs-fmt` on that file

File: pkgs/applications/networking/instant-messengers/psi-plus/default.nix
Comment (SuperSandro2000) [inline pkgs/applications/networking/instant-messengers/psi-plus/default.nix:46]:
```suggestion
assert enablePsiMedia -> enablePlugins;
```
We don't do that anywhere.

File: pkgs/applications/networking/instant-messengers/psi-plus/default.nix
Comment (SuperSandro2000) [inline pkgs/applications/networking/instant-messengers/psi-plus/default.nix:65]:
```suggestion
  ];
```

File: pkgs/applications/networking/instant-messengers/psi-plus/default.nix
Comment (unclechu) [reply pkgs/applications/networking/instant-messengers/psi-plus/default.nix]: @SuperSandro2000 sorry, don’t do what? Validate input arguments types?

File: pkgs/applications/networking/instant-messengers/psi-plus/default.nix
Comment (unclechu) [reply pkgs/applications/networking/instant-messengers/psi-plus/default.nix]: I also removed the argument itself.

**Pull request merged**
